Neovim Universal Healthcare

tjdevries opened the pull request:
# Context

This is in response to #4478. It can be thought of as an extension of #4885.

We will be attempting to create an interface that anyone can hook into to display messages about the health of their plugin, script, feature, etc.
# Outline

The general outline (subject to change) is as follows:
- `nvim/runtime/autoload/health.vim`: Sets up the health checker interface. Provides:
  - `g:health_checkers`: The health checkers dictionary that contains all the available checkers to run. The value of the key determines whether it should be run or not (`v:true` and `v:false`)
    - Also provides a way to find health checkers where applicable.
  - `health#check()`: Some report function that can be called. This will run all the enabled doctors, and compile the results into a nice report (which may initially be printing to a nvim buffer). This can be called by `:CheckHealth<bang>`
  - Report functions - These will be used by people making their own health.vim files. 
    - Initial function:
      - `health#report_start(name)`: A function to begin a specific area of a report.
      - `health#report_info(msg)`: A function to report general info from the checker
      - `health#report_warn(msg, [suggestions])`: A function to report warnings, with optional list of suggestions.
      - `health#report_error(msg, [suggestions])`: A function to report errors, with optional list of suggestions.
- `nvim/runtime/autoload/health/nvim.vim`: This is where the current health.vim file goes. All standard Neovim gotchas / checks will go in this file. It could be split out later quite easily.
- Future plugins / checkers would simply need to place a `$PLUGIN/autoload/health/$PLUGIN.vim` and have a function `health#{plugin_name}#check(bang)` that runs the desired health checks.
  - These could also be made into smaller categories, such as `autoload/health/{plug}/python.vim`, `autoload/health/{plugin}/environment.vim`, etc. to allow for further customiation.
  - ~~They would also need to register their plugin with `health#add_checker({checker_name})`~~
  - Any plugin that has a file in `$PLUGIN/autoload/health/**.vim` will automatically be added and enabled as a `health_checker`.
# Future work
- Improved reporting functions:
  - Allow for different formats of health reporting (txt, markdown, json, etc.). This would be under the same interface, just flipping a different flag.
  - Color coding for reports
- Whatever you can think of! 😄 
# Certain To Do's
- [ ] Update documentation
- [x] Add tests
  - [x] Check `test/functional/plugin/shada_spec.lua`
  - [x] Some sort of functional test for the python configuration
- [ ] Make better to do
# Open Questions
- [x] Best way to determine whether a health checker function exists.
- [x] Should the check exit if `health#report_error` is called, or leave that up to the plugin maintainer?

tjdevries committed:
Terrible first outline

tjdevries committed:
Update so it actually calls function

Uses the same function and idea as before. I have moved the old file out
to the autoload directory, so that it can follow exactly as the other
health checkers will do.

tjdevries committed:
Add an easy way to add checkers to the checkerlist

This could at least abstract that away in case we change how things are
set up. I'm not sure if this is the best way though.

ZyX-I commented:
I think you need to add tests to your TODO, check test/functional/plugin/shada_spec.lua (I point at this file because it contains tests for autoload, syntax, ftplugin and plugin files at once).

Hettomei commented:
> Name of health checkers. We could do doctors

If I understand, a command like `:doctors` ? I do not want 'doctor' because of emacs and 'doctor' mode. 
I don't want any mistake for people using both.

tjdevries commented:
@Hettomei The command would remain to be `:CheckHealth`. I have been trying to think of what to call the individual "health checkers" since that name seems clunky to me. I am not opposed to leaving it that way though.

fmoralesc commented:
(I find the mental image of people mistaking `CheckHealth` with emacs doctor mode very funny)

I think `checkers` is fine as a name.

ZyX-I commented:
“Checkers” is fine, but such things were always used for troubleshooting, and I would suggest moving everything to `autoload/troubleshoot*` is better. But `:Troubleshoot` looks like only slightly better name for a command then `:CheckHealth` (I actually do not like either, but first just sounds bad while second has a reference to a “health” concept which I have not heard of in regards to application setup).

tjdevries commented:
The original reason for using `:CheckHealth` is because that's what the current naming scheme is. I am happy to change it to whatever consensus we can come to here.

justinmk commented:
Health checking is very common in the devops. culture. Command is named CheckHealth for parallel form with existing "check" commands: checktime, etc.

"Troubleshooting" is far too long, too many syllables, and difficult to type e.g. in IRC or Reddit. It's also less generic ; I was hoping CheckHealth can be a general entry point for sanity checks like inspecting VIMRUNTIME.

"healthcheck" is what the checkers can be called.

Hettomei commented:
:HowAreYou
Le 17 juin 2016 20:02, "Justin M. Keyes" <notifications@github.com> a
écrit :

> Health checking is very common in the devops. culture. Command is named
> CheckHealth for parallel form with existing "check" commands: checktime,
> etc.
>
> "Troubleshooting" is far too long, too many syllables, and difficult to
> type e.g. in IRC or Reddit. It's also less generic ; I was hoping
> CheckHealth can be a general entry point for sanity checks like inspecting
> VIMRUNTIME.
>
> "healthcheck" is what the checkers can be called.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/4932#issuecomment-226839159>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AB-5QjOc2C6_FDRI_e-a6mDDN_dP3W6Cks5qMuE-gaJpZM4I3_Lz>
> .
>

wsdjeg commented:
agree with @justinmk.

tjdevries commented:
What does everyone think about something like this?

This is a little more involved than the original proposal, but I don't think it would be very difficult to use. It would take away a lot of the hassle that I see in the current python health checker, worrying about indentation, line wraps, etc.

```vim
""
" Inside of autoload/health/nvim.vim
""
function! health#nvim#check() abort
  " Start diagnosing a certain item
  call health#diagnose('Python 3 configuration')

  call health#symptom_check('General Python Environment')
  call health#add_note('Python 3 is in the path', 'success')
  " Say a symptom check
  call health#symptom_check('Virtual Environment')
  " Add notes about the system
  call health#add_note('Virtual env not configured.', 'warning')
  nvim_pkg_suggestions = [
        \ 'Attempt to install the Neovim pacakage with `pip3 install neovim`',
        \ 'Consult the Neovim Python FAQ page at <website stuff>'
        \ ]
  call health#add_note('Neovim package not installed', 'warning', nvim_pkg_suggestions)
endfunction
```

```
""
" After call of health#check() for this checker
""

" Prints something like this
Nvim Health Checker
  - Diagnosing: Python 3 Configuration
    - General Python Environment
      - SUCCESS: Python 3 is in the Path
    - Virtual Environment
      - WARNING: Virtual env is not configured
      - WARNING: Neovm package not installed
        - SUGGESTION: Attempt to install the Neovim pacakage with `pip3 install neovim`,
        - SUGGESTION: Consult the Neovim Python FAQ page at <website stuff>
```

justinmk commented:
@tjdevries Looks like a nice API. If the names can be made more obvious, it will further help plugin authors.

- `health#diagnose`: let's omit this and just list the function name: `health#nvim#check`. (Can always add this later if it's really needed.)
- `health#symptom_check`: more-obvious name is `health#report_start()` 
- `health#add_note(msg, severity)`: it will help plugin authors if we have separate functions for each severity level (helps with auto-complete, validation, documentation): 
    - `health#report_ok(msg)`
    - `health#report_info(msg)` 
    - `health#report_warn(msg, suggestions)`

```vim
function! health#nvim#check() abort
  call health#report_start('General Python Environment')
  call health#report_ok('Python 3 is in the path')
  call health#report_start('Virtual Environment')
  call health#report_warn('Virtual env not configured.')
  nvim_pkg_suggestions = [
        \ 'Attempt to install the Neovim pacakage with `pip3 install neovim`',
        \ 'Consult the Neovim Python FAQ page at <website stuff>'
        \ ]
  call health#report_warn('Neovim package not installed', nvim_pkg_suggestions)
endfunction
```

tjdevries commented:
@justinmk That seems to be straight forward enough for me. I like it and it seems to be pretty easy to adopt. 

One thing I'm still considering is the levels. Would `ok` be for things that are as expected, `info` be for more of debug output (so environment variables, etc.) and `warn` be for unexpected results? Would we need any other levels? I suppose we could add them later.

justinmk commented:
`ok` is "success"; `info` is just informative messages. What `warn` means, and whether we need another level, I was going to leave up to you.

tjdevries commented:
`ok` ;)

I will think about it some more.

tjdevries committed:
Intermediate commit. Will be rebased

tweekmonster commented:
> whether we need another level, I was going to leave up to you

You're talking about `health#report_wtf()` right? 😉 

But seriously, I think `error` would be needed to report on things that are obviously broken causing further diagnostics to be pointless/impossible.

Commit:
Write some basic implementation for reports

Commit:
Begin to update the original python doctor

I'm hoping to use this as a good starting test for how to use the new
functions. I'm not sure exactly what will happen here or if I'm doing it
all correctly as of yet.

tjdevries committed:
Continue updating the original health file

Also adding various features from that file that seem to be reusable.
I'm still not sure about the notes item, or what should be done with
that. Maybe just a list of compiled notes that could go together at
thend.

tjdevries commented:
Examples of some output that I've gotten from the new setup:

```Checking health
Checker health#nvim#check says: true
  - Diagnosing: Python 2 Configuration
    - INFO: Using: g:python_host_prog = "/usr/bin/python"
    - INFO: Executable:/usr/bin/python
    - INFO: Python Version: 2.7.6
    - INFO: python-neovim Version: 0.1.8
    - WARNING: Latest Neovim Python client versions: 0.1.9)

  - Diagnosing: Python 3 Configuration
    - INFO: Using: g:python3_host_prog = "/usr/bin/python3"
    - INFO: Executable:/usr/bin/python3
    - INFO: Python Version: 3.4.3
    - INFO: python3-neovim Version: 0.1.8
    - WARNING: Latest Neovim Python client versions: 0.1.9)

  - Diagnosing: Remote Plugins
    - INFO: Up to date
```

```Checking health
Checker health#nvim#check says: true
  - Diagnosing: Python 2 Configuration
    - INFO: Using: g:python_host_prog = "/usr/bin/python"
    - INFO: Executable:/usr/bin/python
    - INFO: Python Version: 2.7.6
    - INFO: python-neovim Version: 0.1.8
    - WARNING: Latest Neovim Python client versions: 0.1.9)

  - Diagnosing: Python 3 Configuration
    - INFO: Using: g:python3_host_prog = "/usr/bin/python3"
    - INFO: Executable:/usr/bin/python3
    - INFO: Python Version: 3.4.3
    - INFO: python3-neovim Version: 0.1.8
    - WARNING: Latest Neovim Python client versions: 0.1.9)

  - Diagnosing: Remote Plugins
    - INFO: Up to date
```

tjdevries committed:
Change some of the info calls to success.

tweekmonster commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,408 @@
+      endif
+
+      if resolve(python_bin) !~# '^'.venv_root.'/'
+        call health#report_warn('Your virtualenv is not set up optimatlly.', [
```
`optimally` is spelled wrong here.

tweekmonster commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,408 @@
+            \ ])
+    endif
+
+    if !empty(venv) && exists('g:'.host_prog_var)
```
I think I made a mistake and it should be `elseif !empty(venv) && !exists('g:'.host_prog_var)`.  I'm running on fumes and need a second opinion here.

tjdevries replied:
Yes, that makes sense, otherwise it will run again, regardless.

tjdevries replied:
I'm still not sure which way it should be. @tweekmonster did you come to a conclusion?

tweekmonster replied:
Yeah I'm pretty sure it should be `elseif !empty(venv) && !exists('g:'.host_prog_var)` since we're looking for a setup that's inferred.  The `if` statement above it should have `empty('venv') && ` as well.

The `pyenv` check is to see if the found python interpreter *is not* within `pyenv`'s managed paths.  `virtualenv` is suggested since I find it prudent to keep the base installations pristine.  I think I got confused because the message mentions `virtualenv`.  The message should be `pyenv is not setup optimally`.

I think it got janky because I was moving bits around.  If you don't mind, I can submit a PR to fix this portion.

tjdevries replied:
Yeah, how would that PR work? Put whatever you think is the right logic. I tried not to touch any logic throughout the whole file, so I would love to have you be the person attached to that in the history.

tweekmonster replied:
I imagine I just submit the PR to your fork.  I've never done this before, so if all else fails, I'll send you a patch 😄

tweekmonster commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,408 @@
+
+
+function! health#nvim#check() abort
+  redir => report
```
Shouldn't this be moved to `health#check()`?

tjdevries replied:
Which part exactly?

tweekmonster replied:
The `redir` capturing part since that should be the only point of failure for output.  Leaving that up to checker might cause output leaking out of order or an exception thrown without `redir END`

Shougo replied:
It is better to use `redir()` instead of `:redir`.

tweekmonster replied:
@Shougo What do you mean?  I can't find a `redir()` function.

justinmk replied:
He meant `capture()`

tweekmonster replied:
Ah, didn't know it existed.  Looks like a list of checkers could be built up and ran through `capture()`

Shougo replied:
Yes, it is `capture()`. Sorry.

tjdevries replied:
Changed it so that you don't have to worry about the capture / redit problem in each checker. It's handled in `health.vim`. Thanks for the tip.

tjdevries committed:
Spelling fix

tweekmonster commented:
Responding to your commit:
> I'm still not sure about the notes item, or what should be done with that.

I used it as a way to move the messages below what I thought were the prime diagnostic information, such as version numbers and paths.  Things that could be a dead give away to a problem.

I think how your examples look now is fine, but the `provider#pythonx#Detect()` output can get pretty verbose.  Maybe it'll be less of an issue if message labels were colored to make it easier to distinguish the type of message?

tjdevries committed:
Use capture() instead of redir

tjdevries commented:
Sure, color coding would be pretty cool! I think maybe that will be one of the features to add in future work, for nicer formatting. Adding it to the description above.

tweekmonster committed:
Fixed logic on suboptimal pyenv/virtualenv checks.

tjdevries committed:
Remove some deprecated calls

justinmk commented:
> Best way to determine whether a health checker function exists.

https://github.com/Shougo/unite.vim has a way of doing it. @Shougo how does unit check for unite sources implemented by plugins? Can you link to the relevant code?

> Should the check exit if health#report_error is called, or leave that up to the plugin maintainer?

`report_error` is only a cosmetic effect in the report, it should not cause an actual error. Control flow is totally update to the plugin author.

tweekmonster commented on runtime/autoload/health.vim:
```diff
@@ -1,141 +1,135 @@
+      let s:current_checker = l:checker[0]
+      let l:report .= printf("\nChecker %s says: %s", s:current_checker, l:checker[1])
 
+      let l:report .= capture('call ' . l:checker[0] . '()')
```
Should this be put in a `try` block to catch unforeseen exceptions or will `capture()` grab the exception output?

justinmk replied:
See https://github.com/neovim/neovim/pull/4697#issuecomment-227634187

- `capture('foo')` will capture the exception _and_ display it to the user
    - can be wrapped in `try` if you need to handle the exception in some way
- `capture('silent! foo')` will capture the exception without showing it to the user

tweekmonster replied:
Thanks!  Then, it would only make sense to use `try` if we want a checker to have an exit code (I can't think of a reason where we would).  Otherwise, I think `capture('silent! foo')` would be the best way to go.

Shougo commented:
> https://github.com/Shougo/unite.vim has a way of doing it. @Shougo how does unite check for unite sources implemented by third-party plugins? Can you link to the relevant code?

This.
https://github.com/Shougo/unite.vim/blob/master/autoload/unite/init.vim#L545

unite.vim just search from `'runtimepath'`.

Shougo commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,393 @@
+function! s:download(url) abort
+  let content = ''
+  if executable('curl')
+    let content = system('curl -sL "'.a:url.'"')
```
I think using list is better like this `system(['curl', '-sL', a:url])`.

Shougo commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,393 @@
+          \except Exception:\n
+          \    pass\n
+          \"
+    let content = system('python -c "'.script.'" 2>/dev/null')
```
I think using list is better.

Shougo commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,393 @@
+" Neovim's Python client was updated.
+function! s:version_info(python) abort
+  let pypi_version = s:latest_pypi_version()
+  let python_version = s:trim(system(
```
I think using list is better

Shougo commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,393 @@
+
+  if exists('$VIRTUAL_ENV')
+    if !empty(pyenv)
+      let pyenv_prefix = resolve(s:trim(system(printf('"%s" prefix', pyenv))))
```
I think using list is better

brcolow commented on runtime/autoload/health.vim:
```diff
@@ -145,289 +139,56 @@ function! s:echo_notes(notes) abort
-    for path_bin in python_multiple
-      echo '     (other):' path_bin
+""
+" health#add_checker is a function to register a (or several) healthcheckers.
```
Would it be better to replace "register a (or several)" with "register at least one"? Just a thought.

tjdevries committed:
Use a silent call for the capture

Also fix a small grammer mistake in a comment

tjdevries committed:
Make an attempt at highlighting

tjdevries committed:
Move system calls from strings to lists

Also started trying to give more descriptive output when something
happens that is unexpected, instead of just saying unknown. This
is not only to help me figure out where things are going wrong,
but should also allow other people to get more specific information.

tjdevries committed:
Change a string to a list that I missed

tjdevries committed:
Add some basic highlighting

tjdevries committed:
Add small amount more of highlighting

Next on to try and automatically add health checkers to the
dictionary if they have been made.

tjdevries committed:
Formatting

tjdevries committed:
Add another syntax highlighting rule

tjdevries committed:
Add automatic checker (like Unite)

Thanks Shougo!

Also, added a nicer way of printing messages. Continuing to add
abstraction here so we can add more formats later. Tried to prett
up the output a little as well.

tjdevries commented:
Current output (I'm just leaving in markdown, cause it looks pretty that way :) )

# Checking health

--------------------------------------------------------------------------------

Checker health#gruvbox_tj#check says:

  - Diagnosing: Gruvbox for TJ
    - SUCCESS: All is well
    - WARNING: This is not as good
      - SUGGESTION: This thing
      - SUGGESTION: And the other thing!

--------------------------------------------------------------------------------

Checker health#nvim#check says:

  - Diagnosing: Python 2 Configuration
    - INFO: Using: g:python_host_prog = "/usr/bin/python"
    - WARNING: No Python interpreter was found with the neovim module.  Using the first available for diagnostics.
    - WARNING: provider/pythonx: Could not load Python 2 from g:python_host_prog: /usr/bin/python does not have the neovim module installed. See ":help provider-python".
    - INFO: Executable:/usr/bin/python
    - WARNING: Got an unexpected version of Python. This could lead to confusing error messages.
    - INFO: Python Version: unknown
    - INFO: python-neovim Version: not found
    - ERROR: Neovim Python client is not installed.
      - SUGGESTION: Use the command `pip2install neovim`
    - SUCCESS: Latest Neovim Python client is installed: (unable to get neovim executable)

  - Diagnosing: Python 3 Configuration
    - INFO: Using: g:python3_host_prog = "/usr/bin/python3"
    - INFO: Executable:/usr/bin/python3
    - WARNING: Got an unexpected version of Python. This could lead to confusing error messages.
    - WARNING: Python 3.3+ is recommended.
    - INFO: Python Version: unknown
    - INFO: python3-neovim Version: not found
    - ERROR: Neovim Python client is not installed.
      - SUGGESTION: Use the command `pip3install neovim`
    - SUCCESS: Latest Neovim Python client is installed: (unable to get neovim executable)

  - Diagnosing: Remote Plugins
    - SUCCESS: Up to date

--------------------------------------------------------------------------------

End of output. 

So I implemented something similar at least to what @Shougo did in Unite to find the checkers. It adds them to the dictionary of checkers and automatically runs them (I tested it out in a directory on my comp). 

It's also a little better formatted now. I've added some basic syntax highlighting that could be improved later to help people when they're scanning the output in nvim.

tjdevries committed:
Temporary commit to hold pass all args

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,410 @@
+" Script variables
+let s:bad_responses = ['unknown', 'unable to parse', 'unable to get pypi response', 'unable to get neovim executable']
```
is this still needed?

tjdevries replied:
I haven't finished that part. I was hoping to remove all of the `unknown` calls from the script, and replace them with more verbose names. Then I would check if the response was in the responses, and if so, then it would error. Currently we just check if it's `unknown` in a lot of places.

So it's not needed yet, but I hope to use it soon.

justinmk replied:
I would suggest doing that change in a separate PR.

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -145,289 +156,70 @@ function! s:echo_notes(notes) abort
+" {checker_name} can be specified by either a list of strings or a single string.
+" The string should be the name of the function to check, which should follow
+" the naming convention of `health#plugin_name#check`
+function! health#add_checker(checker_name) abort
```
This should not be public anymore, if it's needed at all.

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,410 @@
+
+
+" Load the remote plugin manifest file and check for unregistered plugins
+function! s:diagnose_manifest() abort
```
"diagnose" should be replaced with "check". Avoid unnecessary jargon.

tjdevries replied:
I think I have removed them all now.

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,410 @@
+    endif
+  endif
+
+  call health#report_notes(notes)
```
Does the checker itself need to call this function? It should be done implicitly by `CheckHealth foo`, not the checker itself.

Then `health#report_notes` can be made private (`s:echo_report()` or whatever).

tjdevries replied:
It was left over from before essentially. I think this is covered by using `health#report_info` now, or by warning with a list of warnings. I removed it and the original function definition.

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -0,0 +1,410 @@
+endfunction
+
+
+function! s:diagnose_python(version) abort
```
`s:check_python`

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -145,289 +156,70 @@ function! s:echo_notes(notes) abort
-    if latest == 'unknown'
-      call add(notes, 'Warning: Unable to fetch latest Neovim Python client version.')
-    endif
+function! health#enable_checker(checker_name) abort
```
This can be private (`s:enable_checker()`). Or even inlined...

tjdevries replied:
Inlined and removed.

tjdevries committed:
Clean up

Cleaned up several functions that no longer need to be public facing, as
mentioned by justinmk.

Also removed any references to notes, as they really don't fit in with
the general idea of the feature otherwise. Report that something
succeeded or that it did not. One can use info if required.

Removed any references to `diagnose`

tjdevries committed:
Remove Diagnosing

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
-    for path_bin in python_multiple
-      echo '     (other):' path_bin
+""
+" s:add_checker is a function to register at least one healthcheckers.
```
The phrase "s:add_checker is a function to" is not necessary. In general, comments don't need to introduce the thing they're describing; it's obvious that this is describe the `s:add_checker` thing, so it doesn't need to be said. Instead, this is the only comment that is needed:

    " Registers at least one healthchecker.

This advice applies to all comments, there are other cases below and above.

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
-          \ . 'using Zsh, see: http://vi.stackexchange.com/a/7654/5229')
+""
+" s:add_single_checker is a function to handle adding a checker of name
+" {checker_name} to the list of health_checkers. It also enables it.
```
" Registers and enables a checker.

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
+      " Report each suggestion
+      for l:suggestion in a:1
+        let l:output .= "\n      - SUGGESTION: " . l:suggestion
+      endfor
```
If we're already forcing the caller to send the suggestions as a list, we can avoid prepending `SUGGESTION` to each line. Instead, it could be formatted like this:

```
SUGGESTIONS:
- {suggestion 1...}
- {suggestion 2...}
- {suggestion 3...}
```

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
+  " Check optional parameters
+  if a:0 > 0
+    " Suggestions go in the first optional parameter and must be a list.
+    if type(a:1) == type([])
```
If we require it to be a list, rather than silently ignoring non-list input, throw an error or else convert it to a string (`string(a:1)`).

tjdevries replied:
Okay, I added something in for this. Could you mention if that's good enough?

justinmk replied:
Looks good. Looking forward to merging this!

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
-    call add(notes, 'Run :UpdateRemotePlugins')
+""
+" This function represents a check that has failed and because of it does not
+" make sense to continue testing the health of the health checker.
```
I'm confused about the mention of "does not make sense to continue". Why doesn't it make sense? Why can't a checker report multiple errors?

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
+  return substitute(substitute(substitute(a:name, ".*autoload/", "", ""), "\\.vim", "#check", ""), "/", "#", "g")
+endfunction " }}}
 
+function! health#_default_checkers() abort " {{{
```
should be private (`s:get_default_checkers()`)

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -1,433 +1,181 @@
+function! health#_default_checkers() abort " {{{
+  " Get all of the files that are in autoload/health/ folders with a vim
+  " suffix
+  let checker_files = globpath(&runtimepath, 'autoload/health/*.vim', 0, 1)
```
We definitely want the `{nosuf}` parameter to be `1`, not `0`--otherwise `'wildignore'` et. al. applies to the search, which makes it dependent on the user's configuration.

tjdevries committed:
Continued cleanup

tjdevries committed:
Allow for strings and throw an error on suggestion

tjdevries committed:
Change {nosuf} parameter

justinmk commented on runtime/autoload/health.vim:
```diff
@@ -70,12 +70,18 @@ function! s:format_report_message(status, msg, ...) abort " {{{
+      let l:output .= "\n        - " . a:1
+    elseif type(a:1) == type([])
       " Report each suggestion
+      let l:output .= "\n      - SUGGESTIONS:
```
looks like a closing quote is missing

tjdevries commented:
@justinmk where would you suggest putting tests for this feature? And what would you suggest testing?

tjdevries committed:
Fix missing parenth... :'(

justinmk commented:
@tjdevries Just a few sanity checks is enough IMO.  `test/functional/plugin/health_spec.lua` seems the natural place.

tjdevries commented:
@justinmk I've been a little busy lately (by that I mean distracted by other neovim things :) ), but I haven't really been able to think of good tests for this, Most of the functions we have are private. so I'm not sure exactly how to call them in the test to see if they update things correctly,

Do you have any suggestions or anything in particular that you'd like to see? I do want to try and get this finished up so people can start having some ideas about implementing their own debug tools for plugins.

justinmk commented:
Just 1 test is needed. In the test, perform the steps required to set up https://github.com/neovim/neovim/pull/4932#issuecomment-229437777 , then call `:CheckHealth!`, and use `helpers.expect()` to verify the buffer contents.

tjdevries commented:
Would it be acceptable to just expect certain phrases or something like that? Obviously some of the environment things would be difficult to expect based on where people run the tests.

justinmk commented:
Oh, for sure. No need to run any actual checks. We're only testing the structure of the output.

tjdevries committed:
Fix incorrect system calls

I also just added in some basic notes about what each failing was.
This was mostly to help me find where the system calls were failing,
since I could not figure out why there doing that initially.

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -3,9 +3,20 @@ let s:bad_responses = [
+        return v:true
+    else
+        return v:false
+    endif
```
just 

    index(s:bad_responses, a:s) >= 0

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -3,9 +3,20 @@ let s:bad_responses = [
             \ 'unable to get pypi response',
-            \ 'unable to get neovim executable'
+            \ 'unable to get neovim executable',
+            \ 'unable to find neovim version'
```
if this enhancement is still WIP it would be better to leave it for another PR.

tjdevries replied:
I think it's good now, The problem was that when I changed things to using the list, as Shougo mentioned I should, for system calls, it broke several calls, but I couldn't figure out where. So while I was debugging, I just added this, and then figured it'd probably be more helpful to just leave it. I can try and remove all of it, but it seems like in the end that just takes more work. I will make the change you mentioned above,

tjdevries replied:
For example, there were times when something could be set to `unknown` in multiple places. You might say `let neovim_version = "unknown"`, and then check stuff, and then get past a set of checks, and then set it again. I just made it so the error was explicit what coulnd't be found.

tjdevries replied:
Or there were cases where we returned something like this `[neovim_version, "unknown", other_var]` and it wouldn't be clear what the `unknown` meant.

justinmk replied:
ok

tjdevries committed:
Simplify function call

tjdevries committed:
Nicer markdown output

tjdevries commented:
@justink, where would I write the help documentation for this feature?

Also, could you point me to a test where it executes a command and then checks for patterns in the buffer? I'm having some trouble with finding that.

tjdevries committed:
Small cleanup and delete unused comment

tjdevries committed:
Fix an inconsistency that for pkg info location

justinmk commented:
> where would I write the help documentation for this feature?

@tjdevries I suggest a new help file `runtime/doc/pi_health.txt`. This is the convention for standard plugins.

tjdevries committed:
Temp commit.

tjdevries committed:
Add tests

tjdevries committed:
Begin writing help documentation

Very rough

tjdevries committed:
Move old reference to checkhealth

I moved it to the new location of the documentation. Now the tags point
to the correct place.

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+Health.vim aims to solve this problem in two ways for both core and plugin
+maintainers.
+
+The way this is done is to provide an interface for that users will know to
```
remove extraneous "for"

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+maintainers.
+
+The way this is done is to provide an interface for that users will know to
+check first before posting questsion in the issue tracker, dev list, etc. This
```
typo in "questsion". Also...instead of "dev list" maybe say gitter or chatroom?

tweekmonster replied:
I think "gitter" might be a little too specific.  Maybe "dev channels" as a broader term.

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+The way this is done is to provide an interface for that users will know to
+check first before posting questsion in the issue tracker, dev list, etc. This
+is similar to how |:help| functions currently. The user experiencing
+difficulty can run |:CheckHealth| to view the status of her system.
```
Just wondering why the "her" pronoun? If we want to go PC why not just use "their"?

jamessan replied:
A pretty common style is to alternate male/female pronouns across paragraphs.  However, the usage of "her" here does standout since most of the other text uses terminology like "the user".  It'd be good to be consistent one way or the other.

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+
+The first aim is to provide maintainers with an overview of the state of the
+user's system. This skips large amounts of time where the maintainer must
+instruct the user what steps to take to get debug information, and allows the
```
"what" -> "which" (IMHO)

tweekmonster replied:
Might read better as: "instruct the user *on* what steps to take".  I think "which" is interchangeable with "what" in this case.

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+is found.
+
+The second aim is to provide maintainers a way of automating the answering of
+frequently encountered questsion. A common occurence with Neovim is that the
```
typo: "questsion" -> "questions"

brcolow commented on runtime/doc/pi_health.txt:
```diff
@@ -3,16 +3,64 @@
+The second aim is to provide maintainers a way of automating the answering of
+frequently encountered questsion. A common occurence with Neovim is that the
+user has not installed the necessary Python modules to interact with Python
+remote plugins.  So a simple check of whhether the neovim remote plugin is
```
typo "whhether" -> "whether". Also consider removing the first word of the sentence "So"

tweekmonster commented on runtime/doc/pi_health.txt:
```diff
@@ -0,0 +1,73 @@
+
+The aim of |:CheckHealth| is two-fold.
+
+The first aim is to provide maintainers with an overview of the state of the
```
"overview of the state of the user's system" is a mouth full.  Could be reworded "overview of the user's system state" or "overview of the user's working environment".

tjdevries committed:
Small alignment adjustment

justinmk commented:
Let's not get tied up in the help doc, I'll tweak before merge. Is this functionally complete?

tjdevries commented:
@justinmk Yes. I think so. The tests that I wrote are working as expected. It seems to find any correctly structured helper that is in the `runtimepath`. The output seems to go nicely straight to markdown, which will be good for pasting into github / other markdown supported places.

If there are any particular concerns, I will try and address those as well, but it seems functional to me.

tjdevries committed:
Move original call back to file

Realized that I messed something up when I was trying to add
something... I'll try and add it in a different PR later. Just
moved the original call back to the function.

Also fixed a systemlist problem

tjdevries committed:
Simplify interface.

The old way would always enable all the checkers when you would
run :CheckHealth, which is bad if people wanted to eliminate some
checkers. So I made the interface consistent.

tjdevries committed:
Add test for disabling checkers

tjdevries committed:
[WIP] Add more documentation

Fixed most of the suggestions that were given.

tjdevries committed:
Finish up my thought on the docs

tjdevries commented:
@justinmk Are you looking for anything else in particular currently? Should I change this to [RFC]?

justinmk commented:
@tjdevries Working on merging this (https://github.com/neovim/neovim/pull/5205). 

I can't find where we discussed the concept of "disabling" a checker. What is the use-case for that?

justinmk commented:
> It's primarily because we add all the checker by default, I figured we should have a way to turn them off if someone desires.

Addressed this in #5205 by letting `:CheckHealth` take optional list of checkers (plugin names) to run.
